content/blog/2019-10-27-tbjit-type-specialization/index.md

```diff
@@ -0,0 +1,189 @@
+
+### Compilation
+
+The LIR traces must first be compiled to machine code to execute natively on the processor. This compilation needs to be much faster than static compilation because it occurs during runtime. The authors propose limiting the number of code optimizations performed to keep the compilation runtime reasonable. For example, register allocation uses a greedy algorithm. Although the authors did not evaluate this, greedy algorithms generally give non-optimal results, but have an excellent accuracy (compared to oracle) per time ratio.
+
+The compiled traces are stored in a trace buffer for later use by the interpreter.
+
+```c
+// LIR code
+guard a == true;
+int c = add_int c 1;
+
+// Machine code
+addi t0 x0 1;
+bne  t1 t0 abort;
+addi t2 t2 1;
+
+```
+
+### Native Execution
+
+The interpreter can execute traces when certain conditions are met. Effectively, the interpreter cedes program control to the generated native machine instructions. The performance of this code should approach that of static code, but is somewhat held back by the low-effort optimizations and additional guard instructions. However, the performance is much better than running in the interpreter.
+
+### Aborting
+
+Whenever a guard fails, we must abort from the current trace because our assumptions were wrong. For example if we thought the type of a value was `int`, but the value turned out to be a `float` future instructions will have incorrect behavior. A simple example is shown below.
+
+```c 
+
+lw  t0 0(s0); // unexpected float!
+add t1 t0 t1; // actually need a fadd instruction!
+
+```
+
+The un-optimized version of this mechanism always jumps backs to the interpreter to decide how to proceed. The interpreter can then record a new trace and start executing machine code from that in future iterations. Effectively, the enumerated steps will repeat in the same order.
+
+
+## Optimizations
+
+The authors lower-level implementation of the ideas described above are the main contributions of this paper. They develop multiple optimizations to make traces less likely to abort. Aborts incur a high performance penalty, so the fewer aborts the faster the user program will run. They also develop techniques to reduce the amount of storage required for the compiled traces. 
+
+#### Typed Traces
+
+Each trace is a basic block that has one entry node and no inner control flow. The interpreter will only enter this basic block if the types of the input variables to the block type check. This is more efficient than entering the trace and immediately aborting because the incoming types were incorrect. In the case of multiple traces, the interpreter has the ability to decide which trace to run based on the input variable types and the trace signature (i.e. the type of each variable as would be given in a C function call).
```

Comment:
```suggestion
Each trace is a basic block that has one entry node and no inner control flow. The interpreter will only enter this basic block if the types of the input variables to the block type check. This is more efficient than entering the trace and immediately aborting because the incoming types were incorrect. In the case of multiple traces, the interpreter has the ability to decide which trace to run based on the input variable types and the trace signature (i.e., the type of each variable as would be given in a C function call).
```
https://capra.cs.cornell.edu/styleguide/#egie